sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/FieldBuilder.java

```diff
@@ -0,0 +1,277 @@
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Stack;
+import java.util.stream.Collectors;
+
+/**
+ * Helper to covert model class to search field {@link Field}.
+ */
+public class FieldBuilder {
+    private static final int MAX_DEPTH = 10000;
+    private static final Map<Class<?>, DataType> SUPPORTED_NONE_PARAMETERIZED_TYPE = new HashMap<>();
+
+    static {
+        SUPPORTED_NONE_PARAMETERIZED_TYPE.put(Integer.class, DataType.EDM_INT32);
+        SUPPORTED_NONE_PARAMETERIZED_TYPE.put(int.class, DataType.EDM_INT32);
+        SUPPORTED_NONE_PARAMETERIZED_TYPE.put(Long.class, DataType.EDM_INT64);
+        SUPPORTED_NONE_PARAMETERIZED_TYPE.put(long.class, DataType.EDM_INT64);
+        SUPPORTED_NONE_PARAMETERIZED_TYPE.put(Double.class, DataType.EDM_DOUBLE);
+        SUPPORTED_NONE_PARAMETERIZED_TYPE.put(double.class, DataType.EDM_DOUBLE);
+        SUPPORTED_NONE_PARAMETERIZED_TYPE.put(Boolean.class, DataType.EDM_BOOLEAN);
+        SUPPORTED_NONE_PARAMETERIZED_TYPE.put(boolean.class, DataType.EDM_BOOLEAN);
+        SUPPORTED_NONE_PARAMETERIZED_TYPE.put(String.class, DataType.EDM_STRING);
+        SUPPORTED_NONE_PARAMETERIZED_TYPE.put(Date.class, DataType.EDM_DATE_TIME_OFFSET);
+        SUPPORTED_NONE_PARAMETERIZED_TYPE.put(OffsetDateTime.class, DataType.EDM_DATE_TIME_OFFSET);
+        SUPPORTED_NONE_PARAMETERIZED_TYPE.put(GeoPoint.class, DataType.EDM_GEOGRAPHY_POINT);
+    }
+
+    private static final List<Class<?>> UNSUPPORTED_TYPES = Arrays.asList(Byte.class,
+        Character.class,
+        char.class,
+        Float.class,
+        float.class,
+        Short.class,
+        short.class);
+
+    private static final Map<Class<?>, Field> CLASS_FIELD_HASH_MAP = new HashMap<>();
+
+    /**
+     * Creates a collection of {@link Field} objects corresponding to the properties of the type supplied.
+     *
+     * @param modelClass The class for which fields will be created, based on its properties.
+     * @param <T> The generic type of the model class.
+     * @return A collection of fields.
+     */
+    public static <T> List<Field> build(Class<T> modelClass) {
+        ClientLogger logger = new ClientLogger(FieldBuilder.class);
+        return build(modelClass, new Stack<>(), logger);
+    }
+
+    /**
+     * Recursive class to build complex data type.
+     *
+     * @param curClass Current class to be built.
+     * @param classChain A class chain from {@code modelClass} to prior of {@code curClass}.
+     * @param logger {@link ClientLogger}.
+     * @return A list of {@link Field} that curClass is built to.
+     */
+    private static List<Field> build(Class<?> curClass, Stack<Class<?>> classChain, ClientLogger logger) {
+        if (classChain.contains(curClass)) {
+            logger.warning(String.format("There is circular dependencies %s, %s", classChain, curClass));
```

Comment:
Do we want to log a warning for a circular dependency? There are a lot of data structures that will have this, for example a binary tree. Would it be better to just return `null` to break the cycle?

Reply:
I have returned null below.
```
if (classChain.contains(curClass)) {
    logger.warning(...);
    return null;
}
```

This is decided by team, which we ignore any property of a type that's already been "seen".

Reply:
No. As discussed with @brjohnstmsft and in other meetings we've had, we want the defaults for model properties and parameters to match the language defaults for those types (e.g. boolean is false). Those should be sent to the service as-is to avoid confusion and force customers to look what the docs currently list as defaults.